src/crypto/qat/qcccrypto.cc

```diff
@@ -322,150 +359,221 @@ bool QccCrypto::perform_op(unsigned char* out, const unsigned char* in,
-    qcc_contig_mem_free((void **)&(qcc_op_mem[avail_inst].src_buff));
-    qcc_op_mem[avail_inst].buff_size = (Cpa32U)size;
-    stat = qcc_contig_mem_alloc((void **)&(qcc_op_mem[avail_inst].src_buff),
-              qcc_op_mem[avail_inst].buff_size);
-    if(stat != CPA_STATUS_SUCCESS) {
-      derr << "Unable to allocate contig memory" << dendl;
-      return false;
+  return status;
+}
+
+CpaStatus QccCrypto::initSession(CpaInstanceHandle cyInstHandle,
+                             CpaCySymSessionCtx *sessionCtx,
+                             Cpa8U *pCipherKey,
+                             CpaCySymCipherDirection cipherDirection) {
+  CpaStatus status = CPA_STATUS_SUCCESS;
+  Cpa32U sessionCtxSize = 0;
+  CpaCySymSessionSetupData sessionSetupData;
+  memset(&sessionSetupData, 0, sizeof(sessionSetupData));
+
+  sessionSetupData.sessionPriority = CPA_CY_PRIORITY_NORMAL;
+  sessionSetupData.symOperation = CPA_CY_SYM_OP_CIPHER;
+  sessionSetupData.cipherSetupData.cipherAlgorithm = CPA_CY_SYM_CIPHER_AES_CBC;
+  sessionSetupData.cipherSetupData.cipherKeyLenInBytes = AES_256_KEY_SIZE;
+  sessionSetupData.cipherSetupData.pCipherKey = pCipherKey;
+  sessionSetupData.cipherSetupData.cipherDirection = cipherDirection;
+
+  if (nullptr == *sessionCtx) {
+    status = cpaCySymDpSessionCtxGetSize(cyInstHandle, &sessionSetupData, &sessionCtxSize);
+    if (likely(CPA_STATUS_SUCCESS == status)) {
+      status = qcc_contig_mem_alloc((void **)(sessionCtx), sessionCtxSize);
+    } else {
+      dout(1) << "cpaCySymDpSessionCtxGetSize failed with status = " << status << dendl;
     }
   }
+  if (likely(CPA_STATUS_SUCCESS == status)) {
+    status = cpaCySymDpInitSession(cyInstHandle,
+                                   &sessionSetupData,
+                                   *sessionCtx);
+    if (unlikely(status != CPA_STATUS_SUCCESS)) {
+      dout(1) << "cpaCySymDpInitSession failed with status = " << status << dendl;
+    }
+  } else {
+    dout(1) << "Session alloc failed with status = " << status << dendl;
+  }
+  return status;
+}
 
-  // Copy src & iv into the respective buffers
-  memcpy(qcc_op_mem[avail_inst].src_buff, in, size);
-  memcpy(qcc_op_mem[avail_inst].iv_buff, iv, AES_256_IV_LEN);
+template <typename CompletionToken>
+auto QatCrypto::async_perform_op(int avail_inst, std::vector<CpaCySymDpOpData*>& pOpDataVec, CompletionToken&& token) {
+  CpaStatus status = CPA_STATUS_SUCCESS;
+  using boost::asio::async_completion;
+  using Signature = void(std::error_code);
+  async_completion<CompletionToken, Signature> init(token);
+  auto ex = boost::asio::get_associated_executor(init.completion_handler);
+  completion_handler = [this, ex, handler = init.completion_handler](std::error_code ec) {
+    boost::asio::post(ex, std::bind(handler, ec));
+  };
+
+  status = cpaCySymDpEnqueueOpBatch(pOpDataVec.size(), &pOpDataVec[0], CPA_TRUE);
+
+  if (status != CPA_STATUS_SUCCESS) {
+    auto ec = std::error_code{status, std::generic_category()};
```

Comment:
> consider using `CpaStatus` instead of `std::error_code` for the `Signature` so we don't have to convert back and forth. `CpaStatus` isn't a generic error code

Done.